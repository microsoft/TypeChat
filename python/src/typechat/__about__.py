# SPDX-FileCopyrightText: Microsoft Corporation
#
# SPDX-License-Identifier: MIT
__version__ = "0.0.2"
