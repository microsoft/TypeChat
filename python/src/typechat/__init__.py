# SPDX-FileCopyrightText: Microsoft Corporation
#
# SPDX-License-Identifier: MIT

from typechat._internal.model import PromptSection, TypeChatLanguageModel, create_language_model, create_openai_language_model, create_azure_openai_language_model
from typechat._internal.result import Failure, Result, Success
from typechat._internal.translator import TypeChatJsonTranslator
from typechat._internal.ts_conversion import python_type_to_typescript_schema
from typechat._internal.validator import TypeChatValidator
from typechat._internal.interactive import process_requests

__all__ = [
    "TypeChatLanguageModel",
    "TypeChatJsonTranslator",
    "TypeChatValidator",
    "Success",
    "Failure",
    "Result",
    "python_type_to_typescript_schema",
    "PromptSection",
    "create_language_model",
    "create_openai_language_model",
    "create_azure_openai_language_model",
    "process_requests",
]
