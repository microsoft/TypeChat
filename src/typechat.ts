import { Result, success, error } from "./result";
import { TypeChatLanguageModel } from "./model";
import { TypeChatJsonValidator, createJsonValidator } from "./validate";

/**
 * Represents an object that can translate natural language requests in JSON objects of the given type.
 */
export interface TypeChatJsonTranslator<T extends object> {
    /**
     * The associated `TypeChatLanguageModel`.
     */
    model: TypeChatLanguageModel;
    /**
     * The associated `TypeChatJsonValidator<T>`.
     */
    validator: TypeChatJsonValidator<T>;
    /**
     * A boolean indicating whether to attempt repairing JSON objects that fail to validate. The default is `true`,
     * but an application can set the property to `false` to disable repair attempts.
     */
    attemptRepair:  boolean;
    /**
     * A boolean indicating whether to delete properties with null values from parsed JSON objects. Some language
     * models (e.g. gpt-3.5-turbo) have a tendency to assign null values to optional properties instead of omitting
     * them. The default for this property is `false`, but an application can set the property to `true` for schemas
     * that don't permit null values.
     */
    stripNulls:  boolean;
    /**
     * Creates an AI language model prompt from the given request. This function is called by `completeAndValidate`
     * to obtain the prompt. An application can assign a new function to provide a different prompt.
     * @param request The natural language request.
     * @returns A prompt that combines the request with the schema and type name of the underlying validator.
     */
    createRequestPrompt(request: string): string;
    /**
     * Creates a repair prompt to append to an original prompt/response in order to repair a JSON object that
     * failed to validate. This function is called by `completeAndValidate` when `attemptRepair` is true and the
     * JSON object produced by the original prompt failed to validate. An application can assign a new function
     * to provide a different repair prompt.
     * @param validationError The error message returned by the validator.
     * @returns A repair prompt constructed from the error message.
     */
    createRepairPrompt(validationError: string): string;
    /**
     * Optionally implements additional validation logic beyond what is expressed in the schema. This function is
     * called following successful schema validation of an instance. By default the function just returns a
     * `Success<T>` for the given instance, but an application can assign a new function that implements any
     * additional validation.
     * @param instance The instance to validate.
     * @returns A `Success<T>` with the final validated instance, or an `Error` explaning the validation failure.
     */
    validateInstance(instance: T): Result<T>;
    /**
     * Translates a natural language request into an object of type `T`. If the JSON object returned by
     * the language model fails to validate and the `attemptRepair` property is `true`, a second
     * attempt to translate the request will be made. The prompt for the second attempt will include the
     * diagnostics produced for the first attempt. This often helps produce a valid instance.
     * @param request The natural language request.
     * @returns A promise for the resulting object.
     */
    translate(request: string): Promise<Result<T>>;
}

/**
 * Creates an object that can translate natural language requests into JSON objects of the given type.
 * The specified type argument `T` must be the same type as `typeName` in the given `schema`. The function
 * creates a `TypeChatJsonValidator<T>` and stores it in the `validator` property of the returned instance.
 * @param model The language model to use for translating requests into JSON.
 * @param schema A string containing the TypeScript source code for the JSON schema.
 * @param typeName The name of the JSON target type in the schema.
 * @returns A `TypeChatJsonTranslator<T>` instance.
 */
export function createJsonTranslator<T extends object>(model: TypeChatLanguageModel, schema: string, typeName: string): TypeChatJsonTranslator<T> {
    const validator = createJsonValidator<T>(schema, typeName);
    const typeChat: TypeChatJsonTranslator<T> = {
        model,
        validator,
        attemptRepair: true,
        stripNulls: false,
        createRequestPrompt,
        createRepairPrompt,
        validateInstance: success,
        translate
    };
    return typeChat;

    function createRequestPrompt(request: string) {
        return `You are a service that translates user requests into JSON objects of type "${validator.typeName}" according to the following TypeScript definitions:\n` +
            `\`\`\`\n${validator.schema}\`\`\`\n` +
            `The following is a user request:\n` +
            `"""\n${request}\n"""\n` +
            `The following is the user request translated into a JSON object with 2 spaces of indentation and no properties with the value undefined:\n`;
    }

    function createRepairPrompt(validationError: string) {
        return `The JSON object is invalid for the following reason:\n` +
            `"""\n${validationError}\n"""\n` +
            `The following is a revised JSON object:\n`;
    }

    async function translate(request: string) {
        let prompt = typeChat.createRequestPrompt(request);
        let attemptRepair = typeChat.attemptRepair;
        while (true) {
            const response = await model.complete(prompt);
            if (!response.success) {
                return response;
            }
            const responseText = response.data;
            const startIndex = responseText.indexOf("{");
            const endIndex = responseText.lastIndexOf("}");
            if (!(startIndex >= 0 && endIndex > startIndex)) {
                return error(`Response is not JSON:\n${responseText}`);
            }
            const jsonText = responseText.slice(startIndex, endIndex + 1);
            const schemaValidation = validator.validate(jsonText);
            const validation = schemaValidation.success ? typeChat.validateInstance(schemaValidation.data) : schemaValidation;
            if (validation.success) {
                return validation;
            }
            if (!attemptRepair) {
                return error(`JSON validation failed: ${validation.message}\n${jsonText}`);
            }
            prompt += `${responseText}\n${typeChat.createRepairPrompt(validation.message)}`;
            attemptRepair = false;
        }
    }
}
