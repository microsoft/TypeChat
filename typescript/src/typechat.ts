import { Result, success, error } from "./result";
import { TypeChatLanguageModel, PromptSection } from "./model";

/**
 * Represents an object that can translate natural language requests in JSON objects of the given type.
 */
export interface TypeChatJsonTranslator<T extends object> {
    /**
     * The associated `TypeChatLanguageModel`.
     */
    model: TypeChatLanguageModel;
    /**
     * The associated `TypeChatJsonValidator<T>`.
     */
    validator: TypeChatJsonValidator<T>;
    /**
     * A boolean indicating whether to attempt repairing JSON objects that fail to validate. The default is `true`,
     * but an application can set the property to `false` to disable repair attempts.
     */
    attemptRepair:  boolean;
    /**
     * A boolean indicating whether to delete properties with null values from parsed JSON objects. Some language
     * models (e.g. gpt-3.5-turbo) have a tendency to assign null values to optional properties instead of omitting
     * them. The default for this property is `false`, but an application can set the property to `true` for schemas
     * that don't permit null values.
     */
    stripNulls:  boolean;
    /**
     * Creates an AI language model prompt from the given request. This function is called by `completeAndValidate`
     * to obtain the prompt. An application can assign a new function to provide a different prompt.
     * @param request The natural language request.
     * @returns A prompt that combines the request with the schema and type name of the underlying validator.
     */
    createRequestPrompt(request: string): string;
    /**
     * Creates a repair prompt to append to an original prompt/response in order to repair a JSON object that
     * failed to validate. This function is called by `completeAndValidate` when `attemptRepair` is true and the
     * JSON object produced by the original prompt failed to validate. An application can assign a new function
     * to provide a different repair prompt.
     * @param validationError The error message returned by the validator.
     * @returns A repair prompt constructed from the error message.
     */
    createRepairPrompt(validationError: string): string;
    /**
     * Optionally implements additional validation logic beyond what is expressed in the schema. This function is
     * called following successful schema validation of an instance. By default the function just returns a
     * `Success<T>` for the given instance, but an application can assign a new function that implements any
     * additional validation.
     * @param instance The instance to validate.
     * @returns A `Success<T>` with the final validated instance, or an `Error` explaning the validation failure.
     */
    validateInstance(instance: T): Result<T>;
    /**
     * Translates a natural language request into an object of type `T`. If the JSON object returned by
     * the language model fails to validate and the `attemptRepair` property is `true`, a second
     * attempt to translate the request will be made. The prompt for the second attempt will include the
     * diagnostics produced for the first attempt. This often helps produce a valid instance.
     * @param request The natural language request.
     * @param promptPreamble An optional string or array of prompt sections to prepend to the generated
     *   prompt. If a string is specified, it is converted into a single "user" role prompt section.
     * @returns A promise for the resulting object.
     */
    translate(request: string, promptPreamble?: string | PromptSection[]): Promise<Result<T>>;
}

/**
 * An object that represents a TypeScript schema for JSON objects.
 */
export interface TypeChatJsonValidator<T extends object> {
    /**
     * Return a string containing TypeScript source code for the validation schema.
     */
    getSchemaText(): string;
    /**
     * Return the name of the JSON object target type in the schema.
     */
    getTypeName(): string;
    /**
     * Validates the given JSON object according to the associated TypeScript schema. Returns a
     * `Success<T>` object containing the JSON object if validation was successful. Otherwise, returns
     * an `Error` object with a `message` property describing the error.
     * @param jsonText The JSON object to validate.
     * @returns The JSON object or an error message.
     */
    validate(jsonObject: object): Result<T>;
}

/**
 * Creates an object that can translate natural language requests into JSON objects of the given type.
 * The specified type argument `T` must be the same type as `typeName` in the given `schema`. The function
 * creates a `TypeChatJsonValidator<T>` and stores it in the `validator` property of the returned instance.
 * @param model The language model to use for translating requests into JSON.
 * @param validator A string containing the TypeScript source code for the JSON schema.
 * @param typeName The name of the JSON target type in the schema.
 * @returns A `TypeChatJsonTranslator<T>` instance.
 */
export function createJsonTranslator<T extends object>(model: TypeChatLanguageModel, validator: TypeChatJsonValidator<T>): TypeChatJsonTranslator<T> {
    const typeChat: TypeChatJsonTranslator<T> = {
        model,
        validator,
        attemptRepair: true,
        stripNulls: false,
        createRequestPrompt,
        createRepairPrompt,
        validateInstance: success,
        translate
    };
    return typeChat;

    function createRequestPrompt(request: string) {
        return `You are a service that translates user requests into JSON objects of type "${validator.getTypeName()}" according to the following TypeScript definitions:\n` +
            `\`\`\`\n${validator.getSchemaText()}\`\`\`\n` +
            `The following is a user request:\n` +
            `"""\n${request}\n"""\n` +
            `The following is the user request translated into a JSON object with 2 spaces of indentation and no properties with the value undefined:\n`;
    }

    function createRepairPrompt(validationError: string) {
        return `The JSON object is invalid for the following reason:\n` +
            `"""\n${validationError}\n"""\n` +
            `The following is a revised JSON object:\n`;
    }

    async function translate(request: string, promptPreamble?: string | PromptSection[]) {
        const preamble: PromptSection[] = typeof promptPreamble === "string" ? [{ role: "user", content: promptPreamble }] : promptPreamble ?? [];
        let prompt: PromptSection[] = [...preamble, { role: "user", content: typeChat.createRequestPrompt(request) }];
        let attemptRepair = typeChat.attemptRepair;
        while (true) {
            const response = await model.complete(prompt);
            if (!response.success) {
                return response;
            }
            const responseText = response.data;
            const startIndex = responseText.indexOf("{");
            const endIndex = responseText.lastIndexOf("}");
            if (!(startIndex >= 0 && endIndex > startIndex)) {
                return error(`Response is not JSON:\n${responseText}`);
            }
            const jsonText = responseText.slice(startIndex, endIndex + 1);
            let jsonObject;
            try {
                jsonObject = JSON.parse(jsonText) as object;
            }
            catch (e) {
                return error(e instanceof SyntaxError ? e.message : "JSON parse error");
            }
            if (typeChat.stripNulls) {
                stripNulls(jsonObject);
            }
            const schemaValidation = validator.validate(jsonObject);
            const validation = schemaValidation.success ? typeChat.validateInstance(schemaValidation.data) : schemaValidation;
            if (validation.success) {
                return validation;
            }
            if (!attemptRepair) {
                return error(`JSON validation failed: ${validation.message}\n${jsonText}`);
            }
            prompt.push({ role: "assistant", content: responseText });
            prompt.push({ role: "user", content: typeChat.createRepairPrompt(validation.message) });
            attemptRepair = false;
        }
    }
}

/**
 * Recursively delete properties with null values from the given object. This function assumes there are no
 * circular references in the object.
 * @param obj The object in which to strip null valued properties.
 */
function stripNulls(obj: any) {
    let keysToDelete: string[] | undefined;
    for (const k in obj) {
        const value = obj[k];
        if (value === null) {
            (keysToDelete ??= []).push(k);
        }
        else {
            if (Array.isArray(value)) {
                if (value.some(x => x === null)) {
                    obj[k] = value.filter(x => x !== null);
                }
            }
            if (typeof value === "object") {
                stripNulls(value);
            }
        }
    }
    if (keysToDelete) {
        for (const k of keysToDelete) {
            delete obj[k];
        }
    }
}
