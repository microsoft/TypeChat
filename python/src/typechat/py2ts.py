from __future__ import annotations

import typing
from dataclasses import dataclass
from types import NoneType, UnionType, get_original_bases
from typing import (
    Annotated,
    Any,
    ClassVar,
    Final,
    Generic,
    Literal,
    LiteralString,
    Never,
    NoReturn,
    NotRequired,
    Protocol,
    Required,
    TypeAlias,
    TypeAliasType,
    TypedDict,
    TypeGuard,
    TypeVar,
    Union,
    cast,
    get_args,
    get_origin,
    get_type_hints,
    is_typeddict,
)

from typechat.ts_type_node import (
    AnyTypeReferenceNode,
    ArrayTypeNode,
    BooleanTypeReferenceNode,
    IdentifierNode,
    IndexSignatureDeclarationNode,
    InterfaceDeclarationNode,
    LiteralTypeNode,
    NeverTypeReferenceNode,
    NullTypeReferenceNode,
    NumberTypeReferenceNode,
    PropertyDeclarationNode,
    StringTypeReferenceNode,
    ThisTypeReferenceNode,
    TopLevelDeclarationNode,
    TypeAliasDeclarationNode,
    TypeNode,
    TypeParameterDeclarationNode,
    TypeReferenceNode,
    UnionTypeNode,
)


class GenericAliasish(Protocol):
    __origin__: type
    __args__: tuple[type, ...]


# type[TypedDict]
# https://github.com/microsoft/pyright/pull/6505#issuecomment-1834431725
class TypeOfTypedDict(Protocol):
    __total__: bool


def is_generic(py_type: object) -> TypeGuard[GenericAliasish]:
    return hasattr(py_type, "__origin__") and hasattr(py_type, "__args__")


TypeReferenceTarget: TypeAlias = type | TypeAliasType | TypeVar | GenericAliasish


def is_python_type_or_alias(origin: object) -> TypeGuard[type | TypeAliasType]:
    return isinstance(origin, TypeAliasType | type)


_KNOWN_GENERIC_SPECIAL_FORMS: frozenset[Any] = frozenset(
    [
        Required,
        NotRequired,
        ClassVar,
        Final,
        Annotated,
        Generic,
    ]
)

_KNOWN_SPECIAL_BASES: frozenset[Any] = frozenset([TypedDict, Protocol])


@dataclass
class TypeScriptTranslationResult:
    type_declarations: list[TopLevelDeclarationNode]
    errors: list[str]


# TODO: https://github.com/microsoft/pyright/issues/6587
_SELF_TYPE = getattr(typing, "Self")

_LIST_TYPES: set[object] = {
    list,
    set,
    frozenset,
    # TODO: https://github.com/microsoft/pyright/issues/6582
    # collections.abc.MutableSequence,
    # collections.abc.Sequence,
    # collections.abc.Set
}

# TODO: https://github.com/microsoft/pyright/issues/6582
# _DICT_TYPES: set[type] = {
#     dict,
#     collections.abc.MutableMapping,
#     collections.abc.Mapping
# }


def pyd_to_ts(root_py_type: object) -> TypeScriptTranslationResult:
    # TODO: handle conflicting names

    declared_types: dict[object, TopLevelDeclarationNode | None] = {}
    undeclared_types = {root_py_type}
    errors: list[str] = []

    def skip_annotations(py_type: object) -> object:
        origin = py_type
        while (origin := get_origin(py_type)) and origin in _KNOWN_GENERIC_SPECIAL_FORMS:
            type_arguments = get_args(py_type)
            if not type_arguments:
                errors.append(f"'{origin}' has been used without any type arguments.")
                return Any
            py_type = type_arguments[0]
            continue
        return py_type

    def convert_to_type_reference_node(py_type: TypeReferenceTarget) -> TypeNode:
        py_type_to_declare = py_type

        if is_generic(py_type):
            py_type_to_declare = get_origin(py_type)

        if py_type_to_declare not in declared_types:
            if is_python_type_or_alias(py_type_to_declare):
                undeclared_types.add(py_type_to_declare)
            elif not isinstance(py_type, TypeVar):
                errors.append(f"Invalid usage of '{py_type}' as a type annotation.")
                return AnyTypeReferenceNode

        if is_generic(py_type):
            return generic_alias_to_type_reference(py_type)

        return TypeReferenceNode(IdentifierNode(py_type.__name__))

    def generic_alias_to_type_reference(py_type: GenericAliasish) -> TypeReferenceNode:
        origin = get_origin(py_type)
        assert origin is not None
        name = origin.__name__
        type_arguments = list(map(convert_to_type_node, get_args(py_type)))
        return TypeReferenceNode(IdentifierNode(name), type_arguments)

    def convert_literal_type_arg_to_type_node(py_type: object) -> TypeNode:
        py_type = skip_annotations(py_type)
        match py_type:
            case str() | int() | float():  # no need to match bool, it's a subclass of int
                return LiteralTypeNode(py_type)
            case None:
                return NullTypeReferenceNode
            case _:
                errors.append(f"'{py_type}' cannot be used as a literal type.")
                return AnyTypeReferenceNode

    def convert_to_type_node(py_type: object) -> TypeNode:
        py_type = skip_annotations(py_type)

        if py_type is str or py_type is LiteralString:
            return StringTypeReferenceNode
        if py_type is int or py_type is float:
            return NumberTypeReferenceNode
        if py_type is bool:
            return BooleanTypeReferenceNode
        if py_type is Any or py_type is object:
            return AnyTypeReferenceNode
        if py_type is None or py_type is NoneType:
            return NullTypeReferenceNode
        if py_type is Never or py_type is NoReturn:
            return NeverTypeReferenceNode
        if py_type is _SELF_TYPE:
            return ThisTypeReferenceNode

        # TODO: consider handling bare 'tuple' (and list, etc.)
        # https://docs.python.org/3/library/typing.html#annotating-tuples
        # Using plain tuple as an annotation is equivalent to using tuple[Any, ...]:

        origin = get_origin(py_type)
        if origin is not None:
            if origin in _LIST_TYPES:
                (type_arg,) = get_type_argument_nodes(py_type, 1, AnyTypeReferenceNode)
                if isinstance(type_arg, UnionTypeNode):
                    return TypeReferenceNode(IdentifierNode("Array"), [type_arg])
                return ArrayTypeNode(type_arg)

            if origin is dict:
                # TODO
                # Currently, we naively assume all dicts are string-keyed
                # unless they're annotated with `int` or `float` (note: not `int | float`).
                key_type_arg, value_type_arg = get_type_argument_nodes(py_type, 2, AnyTypeReferenceNode)
                if key_type_arg is not NumberTypeReferenceNode:
                    key_type_arg = StringTypeReferenceNode
                return TypeReferenceNode(IdentifierNode("Record"), [key_type_arg, value_type_arg])

            # TODO: tuple

            if origin is Union or origin is UnionType:
                type_node = [convert_to_type_node(py_type_arg) for py_type_arg in get_args(py_type)]
                assert len(type_node) > 1
                return UnionTypeNode(type_node)

            if origin is Literal:
                type_node = [convert_literal_type_arg_to_type_node(py_type_arg) for py_type_arg in get_args(py_type)]
                assert len(type_node) >= 1
                return UnionTypeNode(type_node)

            assert is_generic(py_type)
            return convert_to_type_reference_node(py_type)

        if is_python_type_or_alias(py_type):
            return convert_to_type_reference_node(py_type)

        if isinstance(py_type, TypeVar):
            return convert_to_type_reference_node(py_type)

        errors.append(f"'{py_type}' cannot be used as a type annotation.")
        return AnyTypeReferenceNode

    def declare_property(name: str, py_annotation: type | TypeAliasType, optional: bool):
        origin: object = py_annotation
        while origin := get_origin(origin):
            if origin is Required:
                optional = False
                break
            if origin is NotRequired:
                optional = True
                break
        type_annotation = convert_to_type_node(skip_annotations(py_annotation))
        return PropertyDeclarationNode(name, optional, "", type_annotation)

    def declare_type(py_type: object):
        if is_typeddict(py_type):
            assert isinstance(py_type, type)
            type_params = [TypeParameterDeclarationNode(type_param.__name__) for type_param in py_type.__type_params__]
            annotated_members = get_type_hints(py_type, include_extras=True)
            assume_optional = cast(TypeOfTypedDict, py_type).__total__ is False
            raw_but_filtered_bases: list[type] = [
                base
                for base in get_original_bases(py_type)
                if base not in _KNOWN_SPECIAL_BASES and get_origin(base) not in _KNOWN_GENERIC_SPECIAL_FORMS
            ]
            base_properties: dict[str, set[object]] = {}
            for base in raw_but_filtered_bases:
                for prop, annotation in get_type_hints(get_origin(base) or base, include_extras=True).items():
                    base_properties.setdefault(prop, set()).add(annotation)
            properties: list[PropertyDeclarationNode | IndexSignatureDeclarationNode] = [
                declare_property(name, annotation, assume_optional)
                for name, annotation in annotated_members.items()
                # Only keep these in if they're unique or
                if name not in base_properties or
                # all bases declare them differently
                len(base_properties[name]) > 1 or
                # or the current type declares them differently
                annotation not in base_properties[name]
            ]
            bases = [convert_to_type_node(base) for base in raw_but_filtered_bases]
            return InterfaceDeclarationNode(py_type.__name__, type_params, py_type.__doc__ or "", bases, properties)
        if isinstance(py_type, type):
            # TODO: Everything
            errors.append("Currently only TypedDict and type alias declarations are supported")
            return InterfaceDeclarationNode(py_type.__name__, None, f"Comment for {py_type.__name__}.", None, [])
        if isinstance(py_type, TypeAliasType):
            type_params = [TypeParameterDeclarationNode(type_param.__name__) for type_param in py_type.__type_params__]

            return TypeAliasDeclarationNode(
                py_type.__name__,
                type_params,
                f"Comment for {py_type.__name__}.",
                convert_to_type_node(py_type.__value__),
            )

        raise RuntimeError(f"Cannot declare type {py_type}.")

    def get_type_argument_nodes(py_type: object, count: int, default: TypeNode) -> list[TypeNode]:
        py_type_args = get_args(py_type)
        result: list[TypeNode] = []
        if len(py_type_args) != count:
            errors.append(f"Expected '{count}' type arguments for '{py_type}'.")
        for i in range(count):
            if i < len(py_type_args):
                type_node = convert_to_type_node(py_type_args[i])
            else:
                type_node = default
            result.append(type_node)
        return result

    while undeclared_types:
        py_type = undeclared_types.pop()
        declared_types[py_type] = None
        declared_types[py_type] = declare_type(py_type)

    type_declarations = cast(list[TopLevelDeclarationNode], list(declared_types.values()))
    assert None not in type_declarations

    return TypeScriptTranslationResult(type_declarations, errors)
