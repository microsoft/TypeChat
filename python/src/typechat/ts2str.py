import json
from typing import assert_never

from typechat.ts_type_node import (
    ArrayTypeNode,
    IdentifierNode,
    IndexSignatureDeclarationNode,
    InterfaceDeclarationNode,
    LiteralTypeNode,
    NullTypeReferenceNode,
    PropertyDeclarationNode,
    TopLevelDeclarationNode,
    TypeAliasDeclarationNode,
    TypeNode,
    TypeReferenceNode,
    UnionTypeNode,
)


def comment_to_str(comment_text: str, indentation: str) -> str:
    comment_text = comment_text.strip()
    if not comment_text:
        return ""
    lines = [line.strip() for line in comment_text.splitlines()]

    return "\n".join([f"{indentation}// {line}" for line in lines]) + "\n"


def ts_type_to_str(type_node: TypeNode) -> str:
    match type_node:
        case TypeReferenceNode(name, type_arguments):
            assert isinstance(name, IdentifierNode)
            if type_arguments is None:
                return name.text
            return f"{name.text}<{', '.join([ts_type_to_str(arg) for arg in type_arguments])}>"
        case ArrayTypeNode(element_type):
            assert type(element_type) is not UnionTypeNode
            # if type(element_type) is UnionTypeNode:
            #     return f"Array<{ts_type_to_str(element_type)}>"
            return f"{ts_type_to_str(element_type)}[]"
        case UnionTypeNode(types):
            # Remove duplicates, but try to preserve order of types,
            # and put null at the end if it's present.
            str_set: set[str] = set()
            type_strs: list[str] = []
            nullable = False
            for type_node in types:
                if type_node is NullTypeReferenceNode:
                    nullable = True
                    continue
                type_str = ts_type_to_str(type_node)
                if type_str not in str_set:
                    str_set.add(type_str)
                    type_strs.append(type_str)
            if nullable:
                type_strs.append("null")
            return " | ".join(type_strs)
        case LiteralTypeNode(value):
            return json.dumps(value)
        # case _:
        #     raise NotImplementedError(f"Unhandled type {type(type_node)}")
    assert_never(type_node)


def object_member_to_str(member: PropertyDeclarationNode | IndexSignatureDeclarationNode) -> str:
    match member:
        case PropertyDeclarationNode(name, is_optional, comment, annotation):
            comment = comment_to_str(comment, "    ")
            if not name.isalnum():
                name = json.dumps(name)
            return f"{comment}    {name}{'?' if is_optional else ''}: {ts_type_to_str(annotation)};"
        case IndexSignatureDeclarationNode(key_type, value_type):
            return f"[key: {ts_type_to_str(key_type)}]: {ts_type_to_str(value_type)};"
        # case _:
        #     raise NotImplementedError(f"Unhandled member type {type(member)}")
    assert_never(member)


def ts_declaration_to_str(declaration: TopLevelDeclarationNode) -> str:
    match declaration:
        case InterfaceDeclarationNode(name, type_parameters, comment, base_types, members):
            comment = comment_to_str(comment, "")
            type_param_str = f"<{', '.join([param.name for param in type_parameters])}>" if type_parameters else ""
            base_type_str = (
                f" extends {', '.join([ts_type_to_str(base_type) for base_type in base_types])}" if base_types else ""
            )
            members_str = "\n".join([f"{object_member_to_str(member)}" for member in members]) + "\n" if members else ""
            return f"{comment}interface {name}{type_param_str}{base_type_str} {{\n{members_str}}}\n"
        case TypeAliasDeclarationNode(name, type_parameters, comment, target):
            type_param_str = f"<{', '.join([param.name for param in type_parameters])}>" if type_parameters else ""
            return f"type {name}{type_param_str} = {ts_type_to_str(target)}\n"
        # case _:
        #     raise NotImplementedError(f"Unhandled declaration type {type(declaration)}")
    assert_never(declaration)
