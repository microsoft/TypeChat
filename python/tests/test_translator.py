
import asyncio
from dataclasses import dataclass
from typing_extensions import Any, Iterator, Literal, TypedDict, override
import typechat

class ConvoRecord(TypedDict):
    kind: Literal["CLIENT REQUEST", "MODEL RESPONSE"]
    payload: str | list[typechat.PromptSection]

class FixedModel(typechat.TypeChatLanguageModel):
    responses: Iterator[str]
    conversation: list[ConvoRecord]

    "A model which responds with one of a series of responses."
    def __init__(self, responses: list[str]) -> None:
        super().__init__()
        self.responses = iter(responses)
        self.conversation = []

    @override
    async def complete(self, prompt: str | list[typechat.PromptSection]) -> typechat.Result[str]:
        # Capture a snapshot because the translator
        # can choose to pass in the same underlying list.
        if isinstance(prompt, list):
            prompt = prompt.copy()

        self.conversation.append({ "kind": "CLIENT REQUEST", "payload": prompt })
        response = next(self.responses)
        self.conversation.append({ "kind": "MODEL RESPONSE", "payload": response })
        return typechat.Success(response)

@dataclass
class ExampleABC:
    a: str
    b: bool
    c: int

v = typechat.TypeChatValidator(ExampleABC)

def test_translator_with_immediate_pass(snapshot: Any):
    m = FixedModel([
        '{ "a": "hello", "b": true, "c": 1234 }',
    ])
    t = typechat.TypeChatJsonTranslator(m, v, ExampleABC)
    asyncio.run(t.translate("Get me stuff."))
    
    assert m.conversation == snapshot

def test_translator_with_single_failure(snapshot: Any):
    m = FixedModel([
        '{ "a": "hello", "b": true }',
        '{ "a": "hello", "b": true, "c": 1234 }',
    ])
    t = typechat.TypeChatJsonTranslator(m, v, ExampleABC)
    asyncio.run(t.translate("Get me stuff."))
    
    assert m.conversation == snapshot

def test_translator_with_invalid_json(snapshot: Any):
    m = FixedModel([
        '{ "a": "hello" "b": true }',
        '{ "a": "hello" "b": true, "c": 1234 }',
    ])
    t = typechat.TypeChatJsonTranslator(m, v, ExampleABC)
    asyncio.run(t.translate("Get me stuff."))
    
    assert m.conversation == snapshot

def test_translator_with_single_failure_and_str_preamble(snapshot: Any):
    m = FixedModel([
        '{ "a": "hello", "b": true }',
        '{ "a": "hello", "b": true, "c": 1234 }',
    ])
    t = typechat.TypeChatJsonTranslator(m, v, ExampleABC)
    asyncio.run(t.translate(
        "Get me stuff.",
        prompt_preamble="Just so you know, I need some stuff.",
    ))
    
    assert m.conversation == snapshot

def test_translator_with_single_failure_and_list_preamble_1(snapshot: Any):
    m = FixedModel([
        '{ "a": "hello", "b": true }',
        '{ "a": "hello", "b": true, "c": 1234 }',
    ])
    t = typechat.TypeChatJsonTranslator(m, v, ExampleABC)
    asyncio.run(t.translate("Get me stuff.", prompt_preamble=[
        {"role": "user", "content": "Hey, I need some stuff."},
        {"role": "assistant", "content": "Okay, what kind of stuff?"},
    ]))
    
    assert m.conversation == snapshot

