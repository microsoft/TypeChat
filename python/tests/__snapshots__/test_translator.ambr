# serializer version: 1
# name: test_translator_with_immediate_pass
  list([
    dict({
      'kind': 'CLIENT REQUEST',
      'payload': list([
        dict({
          'content': 'Get me stuff.',
          'role': 'user',
        }),
        dict({
          'content': '''
            
            You are a service that translates user requests into JSON objects of type "ExampleABC" according to the following TypeScript definitions:
            ```
            interface ExampleABC {
                a: string;
                b: boolean;
                c: number;
            }
            
            ```
            The following is a user request:
            '''
            Get me stuff.
            '''
            The following is the user request translated into a JSON object with 2 spaces of indentation and no properties with the value undefined:
  
          ''',
          'role': 'user',
        }),
      ]),
    }),
    dict({
      'kind': 'MODEL RESPONSE',
      'payload': '{ "a": "hello", "b": true, "c": 1234 }',
    }),
  ])
# ---
# name: test_translator_with_single_failure
  list([
    dict({
      'kind': 'CLIENT REQUEST',
      'payload': list([
        dict({
          'content': 'Get me stuff.',
          'role': 'user',
        }),
        dict({
          'content': '''
            
            You are a service that translates user requests into JSON objects of type "ExampleABC" according to the following TypeScript definitions:
            ```
            interface ExampleABC {
                a: string;
                b: boolean;
                c: number;
            }
            
            ```
            The following is a user request:
            '''
            Get me stuff.
            '''
            The following is the user request translated into a JSON object with 2 spaces of indentation and no properties with the value undefined:
  
          ''',
          'role': 'user',
        }),
        dict({
          'content': '''
            
            The above JSON object is invalid for the following reason:
            '''
            Validation path `c` failed for value `{"a": "hello", "b": true}` because:
              Field required
            '''
            The following is a revised JSON object:
  
          ''',
          'role': 'user',
        }),
      ]),
    }),
    dict({
      'kind': 'MODEL RESPONSE',
      'payload': '{ "a": "hello", "b": true }',
    }),
    dict({
      'kind': 'CLIENT REQUEST',
      'payload': list([
        dict({
          'content': 'Get me stuff.',
          'role': 'user',
        }),
        dict({
          'content': '''
            
            You are a service that translates user requests into JSON objects of type "ExampleABC" according to the following TypeScript definitions:
            ```
            interface ExampleABC {
                a: string;
                b: boolean;
                c: number;
            }
            
            ```
            The following is a user request:
            '''
            Get me stuff.
            '''
            The following is the user request translated into a JSON object with 2 spaces of indentation and no properties with the value undefined:
  
          ''',
          'role': 'user',
        }),
        dict({
          'content': '''
            
            The above JSON object is invalid for the following reason:
            '''
            Validation path `c` failed for value `{"a": "hello", "b": true}` because:
              Field required
            '''
            The following is a revised JSON object:
  
          ''',
          'role': 'user',
        }),
      ]),
    }),
    dict({
      'kind': 'MODEL RESPONSE',
      'payload': '{ "a": "hello", "b": true, "c": 1234 }',
    }),
  ])
# ---
